src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPullerService.cs

```diff
@@ -20,81 +19,170 @@ internal class EventPullerService : BackgroundService
-                case DomainEventHandlerNotRegisteredException:
-                    this._logger.EventWillBeRejected(cloudEvent.Id, cloudEvent.Type, ex);
-                    await RejectEvent(eventGridTopicSubscription, lockToken, stoppingToken).ConfigureAwait(false);
-                    break;
-                default:
-                    this._logger.EventWillBeReleased(cloudEvent.Id, cloudEvent.Type, ex);
-                    await ReleaseEvent(eventGridTopicSubscription, lockToken, stoppingToken).ConfigureAwait(false);
-                    break;
+                var lockTokens = await ReadCurrentChannelContent(this._releaseEventChannel, cancellationToken).ConfigureAwait(false);
+                await this._eventGridTopicSubscription.Client.ReleaseCloudEventsAsync(this._eventGridTopicSubscription.TopicName, this._eventGridTopicSubscription.SubscriptionName, lockTokens, cancellationToken).ConfigureAwait(false);
             }
         }
-    }
 
-    private static async Task AcknowledgeEvent(EventGridTopicSubscription eventGridTopicSubscription, string lockToken, CancellationToken stoppingToken)
-    { 
-        await eventGridTopicSubscription.Client.AcknowledgeCloudEventAsync(eventGridTopicSubscription.TopicName, eventGridTopicSubscription.SubscriptionName, lockToken, stoppingToken).ConfigureAwait(false);
-    }
+        private async Task RejectEvents(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                var lockTokens = await ReadCurrentChannelContent(this._rejectEventChannel, cancellationToken).ConfigureAwait(false);
+                await this._eventGridTopicSubscription.Client.RejectCloudEventsAsync(this._eventGridTopicSubscription.TopicName, this._eventGridTopicSubscription.SubscriptionName, lockTokens, cancellationToken).ConfigureAwait(false);
+            }
+        }
 
-    private static async Task ReleaseEvent(EventGridTopicSubscription eventGridTopicSubscription, string lockToken, CancellationToken stoppingToken)
-    {
-        await eventGridTopicSubscription.Client.ReleaseCloudEventAsync(eventGridTopicSubscription.TopicName, eventGridTopicSubscription.SubscriptionName, lockToken, stoppingToken).ConfigureAwait(false);
-    }
+        private static async Task<List<string>> ReadCurrentChannelContent(Channel<string> channel, CancellationToken cancellationToken)
+        {
+            await channel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false);
 
-    private static async Task RejectEvent(EventGridTopicSubscription eventGridTopicSubscription, string lockToken, CancellationToken stoppingToken)
-    {
-        await eventGridTopicSubscription.Client.RejectCloudEventAsync(eventGridTopicSubscription.TopicName, eventGridTopicSubscription.SubscriptionName, lockToken, stoppingToken).ConfigureAwait(false);
+            var currentChannelCount = channel.Reader.Count;
+            var results = new List<string>();
```

Comment:
_I don't know if this is important for perf (would be better to test it)._

I think you could preallocate the list in the calling method with the expected capacity. Before calling `ReadCurrentChannelContent`, you can clear it. This way, you only allocates 3 lists for the whole lifetime of the app, instead of one list each time you call this method.

Reply:
The way it currently works, it will grab whatever currently resides within the channel and send it all at once to the acknowledge/release/reject endpoint. There is no way to know how many events are currently sitting in there.

However, it might be more prudent to limit the size of a single batch. There might be a limit on the amount of lockTokens being sent at once, and overall it would allow for reusable lists with fixed sizes, since we know we can't outgrow it. I'll implement this.

Reply:
I changed the way the reading occurs, it will now read up to the current count when starting to read but it returns an `IEnumerable` that is directly passed to the client.

Since the client receives an IEnumerable, I am assuming they will handle proper partitionning of a number of tokens that is too high. This also has the benefit of removing the allocation of intermediary lists.